src/queue.ml

```diff
@@ -269,6 +249,20 @@ module Make (B : Context.A_DISK) = struct
     ; free_sectors = Int64.add free_sectors (Int64.of_int nb)
     }
 
+  let pop_old_generation q last_gen_id =
+    let rec pop queue =
+      let* queue, lst, _ = pop_front queue q.bitset 1 in
+      let range = List.nth lst 0 
+      in
+      if range = last_gen_id
+      then Lwt_result.return queue
+      else
+        let* () = Bitset.free_range q.bitset range in
+        pop queue
+    in
```

Comment:
Yes this works! If you have time, we shouldn't need `pop_front` to pop a specific quantity anymore so it should be possible to specialize it for "pop until the generation separator is encountered" :)